### DISCUSSION TITLE
Crystal orientation issue

### DISCUSSION THREAD

**Discussion by majidkavousi** [neperfepx/neper#434] [Q&A] [answered]
Hi 

I am trying to use Neper for microstructure generation and then use it with Abaqus UMAT crystal plasticity. The problem that I have at the moment is that I don't understand the cell orientations in the .tess file for the Rodrigues case.
![image](https://user-images.githubusercontent.com/99205979/162420251-73d2092d-0177-42a5-8610-7b455bbe39ba.png)
To be clear, for the UMAT, I need two vectors in the crystal coordinate system and their equivalent direction vectors in the global coordinate system. How can I use the *ori values in the tess file for the case of Rodrigues to calculate those vectors?

**Comment by mkasemer** [comment, marked as answer]
@majidmech

The orientations are by default printed as Rodrigues vectors. The orientation output can be printed using various parameterizations, including various flavors of Euler angles and angle-axis parameterizations.

To get what you need, I suggest converting any of the parameterizations printed from Neper to a rotation matrix, and using said matrix to transform two vectors in the sample frame to the crystal frame using standard vector transformation techniques, facilitated through the calculation of the rotation matrix.

**Reply by rquey** [reply]
You can even output the orientations directly as rotation matrices, using [-orides rotmat](https://neper.info/doc/neper_t.html#cmdoption-oridescriptor). The components of the rotation matrix are written in the order 11 12 13 21 22 23 31 32 33.

The vector of components 11 12 13 corresponds to the crystal direction parallel to the sample X direction.
The vector of components 21 22 23 corresponds to the crystal direction parallel to the sample Y direction.
The vector of components 31 32 33 corresponds to the crystal direction parallel to the sample Z direction.

**Reply by mkasemer** [reply]
The above suggestion is, of course, much more direct than what I described.

**Reply by majidkavousi** [reply]
Thanks All. Your suggestion helped me a lot. So I can readily use these rotmat components  for my crystal coordinates.

**Reply by majidkavousi** [reply]
Hi Romain,
I've been working today on this crystal orientation topic and I've tested multiple cases in Neper with only one crystal, to better relate the crystal and sample coordinate systems. Through my tests, I think that maybe what you have told needs to be rephrased as below, correct me if I am wrong:
* The vector of components 11 12 13 corresponds to the **sample direction** parallel to the **crystal X direction**.
* The vector of components 21 22 23 corresponds to the **sample direction** parallel to the **crystal Y direction**.
* The vector of components 31 32 33 corresponds to the **sample direction** parallel to the **crystal X direction**.

Here is the orientations as rotation matrix
*ori
   rotmat:active
 0.835261851153  0.508799068169  0.208473375371 -0.545422340404  0.718640865283  0.431358061631  0.069657092940 -0.474002969374  0.877763678006

I've drawn vectors (11 12 13), (21 22 23) and (31 32 33) against sample coordinates to better visualize the issue:
![image](https://user-images.githubusercontent.com/99205979/162795766-8f3b902b-ba5d-4555-be52-a5323a3ccb1d.png)

**Reply by rquey** [reply]
Yes, I think you are correct, I phrased it wrong.  Anyway, here is what I usually write:

> By construction, its rows contain the coordinates of the crystal coordinate system vectors in the reference coordinate system, and similarly, its columns contain the coordinates of the reference coordinate system vectors in the crystal coordinate system.

**Reply by majidkavousi** [reply]
Thanks for your kind replies. I can go on then.
